feat: StateChart (support for compound / parallel / historical states including SCXML notation)

fgmacedo opened the pull request:
## Statecharts are here

This PR brings **full statechart support** to python-statemachine — the most significant evolution of the library since its creation. What started as an exploration of nested states grew into a comprehensive implementation of the [SCXML specification (W3C)](https://www.w3.org/TR/scxml/), introducing **compound states**, **parallel states**, **history pseudo-states**, **eventless transitions**, **error handling**, and much more.

This is a **major version release (3.0.0)** with backwards-incompatible changes. The `StateMachine` class is preserved for backwards compatibility, while the new `StateChart` class provides full SCXML-compliant behavior out of the box.

### Compound states

States can now contain inner substates using `State.Compound` with Python class syntax:

```python
class ShireToRoad(StateChart):
    class shire(State.Compound):
        bag_end = State(initial=True)
        green_dragon = State()
        visit_pub = bag_end.to(green_dragon)

    road = State(final=True)
    depart = shire.to(road)
```

Entering a compound state activates both the parent and its initial child. Exiting removes the parent and all descendants.

### Parallel states

Multiple regions can be active simultaneously using `State.Parallel`:

```python
class WarOfTheRing(StateChart):
    class war(State.Parallel):
        class frodos_quest(State.Compound):
            shire = State(initial=True)
            mordor = State(final=True)
            journey = shire.to(mordor)
        class aragorns_path(State.Compound):
            ranger = State(initial=True)
            king = State(final=True)
            coronation = ranger.to(king)
```

Events in one region don't affect others — true orthogonal concurrency.

### History pseudo-states

The `HistoryState` records the configuration of a compound state when exited. Re-entering via history restores the previously active child (shallow or deep):

```python
class GollumPersonality(StateChart):
    class personality(State.Compound):
        smeagol = State(initial=True)
        gollum = State()
        h = HistoryState()
        dark_side = smeagol.to(gollum)
    outside = State()
    leave = personality.to(outside)
    return_via_history = outside.to(personality.h)
```

### Eventless (automatic) transitions

Transitions without an event trigger fire automatically when their source state is active, cascading through the entire chain in a single macrostep.

### DoneData on final states

Final states can provide data to `done.state` handlers via the `donedata` parameter, following the SCXML spec for communicating completion results.

### Error handling with `error.execution`

When `error_on_execution` is enabled (default in `StateChart`), runtime exceptions are caught and result in an internal `error.execution` event — following the SCXML error handling specification. A `error_` naming convention makes it easy:

```python
class MyChart(StateChart):
    s1 = State(initial=True)
    error_state = State(final=True)
    go = s1.to(s1, on="bad_action")
    error_execution = s1.to(error_state)  # matches "error.execution" automatically
```

### Additional features

- **Delayed events** — `sm.send("event", delay=500)` with cancellation via `sm.cancel_event(send_id)`
- **Event matching per SCXML spec** — hierarchical event name matching with wildcard `*` support
- **`In(state)` condition checks** — `cond="In('state_id')"` to check active configuration
- **`prepare_event` callback** — inject custom kwargs into all callbacks
- **Multi-target transitions** — a single transition can target multiple states
- **Dictionary-based definitions** — `create_machine_class_from_definition()` for dynamic state machine creation
- **Diagram generation** — Graphviz rendering for compound, parallel, and history states with proper subgraph nesting

### Engine architecture

Both the **sync and async engines** have been rewritten to implement the SCXML processing model:

- 3-phase macrostep loop (eventless → internal queue → external queue)
- Proper enter/exit ordering following the SCXML algorithm
- `configuration` (ordered set of active states) replaces `current_state`
- Internal and external event queues with priority handling

### SCXML compliance

The test suite includes **198 W3C SCXML test cases** (mandatory + optional), run on **both sync and async engines**. Currently **152 pass** and **46 are marked as expected failures** (mostly requiring `<invoke>` which is deferred to v3.1+).

### Quality

- **100% test coverage** — 988 tests, 0 missing lines, 0 missing branches
- Full support for both sync and async engines across all features
- SonarCloud issues addressed
- Comprehensive documentation with doctests

### Backwards compatibility

The `StateMachine` class preserves the old behavior through class-level attribute defaults:

| Attribute | `StateChart` | `StateMachine` |
|---|---|---|
| `allow_event_without_transition` | `True` | `False` |
| `enable_self_transition_entries` | `True` | `False` |
| `atomic_configuration_update` | `False` | `True` |
| `error_on_execution` | `True` | `False` |

Existing `StateMachine` subclasses continue to work as before.

### Known limitations (deferred to v3.1+)

- `<invoke>` — invoking external services or sub-machines
- HTTP and other external communication targets
- `<finalize>` — processing data returned from invoked services

---

Closes #439, #455. Supersedes #329.

fgmacedo committed:
feat: Basic support for SCXML test suit

fgmacedo committed:
feat: Support for onentry/raise/assign/log/if/elseif/else elements; When entering a top level final state the SM terminates

fgmacedo committed:
fix: SCXML Parsing of if/else element

fgmacedo committed:
feat: Support for SCXML foreach tag

fgmacedo committed:
fix: Trigger eventless transition at startup

fgmacedo committed:
refac: Preserving original scxml file names for passing testcases

fgmacedo committed:
fix: Allowing event without transition

fgmacedo committed:
feat: Delayed events; Support for SCXML <send> tag

fgmacedo committed:
refac: Attempt to use PriorityQueue

fgmacedo committed:
refac: Using get/set state for pickling

fgmacedo committed:
feat: Support for SCXML <cancel> tag. Allow cancelling delayed events

fgmacedo committed:
feat: Support for SCXML <script>, all W3C tests imported and ~50% passing

fgmacedo committed:
feat: Support for SCXML <_event.name>

fgmacedo committed:
chore: Drop support for Python3.7 and 3.8 due to the lack of support for TypedDict and | operator

fgmacedo committed:
fix: Fix merge with develop

fgmacedo committed:
refac: Split the SCXML parser into schema definition, parser and actions for executable content

fgmacedo committed:
fix: Fix type hints for Python 3.9 as the | operator is not supported

fgmacedo committed:
feat: Add syntax for compound and parallel states (only parser)

fgmacedo committed:
chore: Remove support for non-RTC model; Preparing processing_loop for new version

fgmacedo committed:
chore: Fix cmd params for test coverage

fgmacedo committed:
feat: SCXML processing model

fgmacedo committed:
fix: Fix compatibility with py3.9

fgmacedo committed:
feat: Hierarquical statemachines with compose and parallel

fgmacedo committed:
feat: Internal transitions

fgmacedo committed:
chore: New SCXML fail mark with the contents of the errors

fgmacedo committed:
chore: Microwave example with parallel state working

fgmacedo committed:
feat: New  callback

fgmacedo committed:
docs: Release example of create_machine_class_from_definition

fgmacedo committed:
fix: Fix parser of sub-final states

fgmacedo committed:
fix: Fix parallel enter/exit and  checks

fgmacedo committed:
chore: SCXML failing xfail marks updated

fgmacedo committed:
fix: Executing <initial> content on default entry

fgmacedo committed:
chore: Fix pyright complaining about event calls

fgmacedo committed:
fix: Fix test533, ordering of entering states and exit/enter of parallel states

fgmacedo committed:
fix: Better support for targetless transitions

fgmacedo committed:
fix: Fix sm name from scxml element

fgmacedo committed:
fix: Fix 326, _ioprocessor is the same for the session

fgmacedo committed:
fix: Fix some edge cases of initial state configuration

fgmacedo committed:
fix: Log may contain only the 'label' attr

fgmacedo committed:
fix: Reading from file on <scxml.datamodel.data.src> tag

fgmacedo committed:
fix: Initials of scxml is respected; Added validate_disconnected_states parameter since scxml don't expect that the graph should have a single component

fgmacedo committed:
fix: SCXML _event should be bound only after the first event; the instance should keep the same

fgmacedo committed:
fix: Fix event tests on the SCXML suit

fgmacedo committed:
feat: Support for History pseudo state

comalice commented:
Is this in a state (🤣) that I could try it out? I have several use-cases where the sub-state functionality would be useful and I'd be glad to give feedback if I run into issues.

fgmacedo commented:
@comalice Thanks for your interest on this. I think the internals are usable, but the external API may still change as I've not yet finished documenting and exploring examples using the Python API.

There are many new *.scxml files that I've used to fine tune the internal behaviour for composite and parallel states, and the `create_machine_class_from_definition` is more stable and mature as a way to construct the statemachine from a dict-like definition similar to the way SCXML works. I'm planning to also add a YAML parser.

The only notable non implemented feature is the `<invoke>` and thus cross sending events between distinct statemachines.

> Note, the user experience I had expected looked something like this:
> 
> ```python
> class SubMachine(StateMachine):
>     test = State(initial=True)
>     test_complete = State(final=True)
> 
>     next = test.to(test_complete)
> 
> 
> class TestSM(StateMachine):
>     initial_state = State(initial=True)
>     sub_machine_as_state = SubMachine()
>     some_other_state = State(final=True)
> 
>     goto_sub_machine = initial_state.to(sub_machine_as_state)
>     goto_finish = sub_machine_as_state.to(some_other_state)
> ```
> 

I may try to implement this syntax, but I think that will be hard to get it right. The initial proposal that I'm implement is the one that is in the description, using the internal namespace of nested `Parallel` and `Compose` classes to build hierarchy.

Feel free to fork and test, please let me know if you expecience any issues. 

Since this got into a huge PR and almost stable, I'm considering merging and finish the work with smaller PRs.

fgmacedo committed:
chore: Declaring new base StateChart with new defaults and keeping StateMachine for backwards compatibility

fgmacedo committed:
chore: merge develop into macedo/scxml

Resolve conflicts from Python 3.14 support merge (PR #552):
- CI matrix: keep scxml's min Python 3.9, add 3.14
- pyproject.toml: add 3.14 classifier, update ruff ignore list
- statemachine.py: use `is not None` check, keep history_values and
  id-keyed listeners from scxml branch
- docs/guards.md: keep expanded declaration order docs from develop
- Fix trailing whitespace and unused imports from scxml branch

fgmacedo committed:
fix: fix syntax error in scxml actions and update diagram doctest

- Parenthesize multiple exception types in except clause (Python 3 syntax)
- Update diagram doctest to match new compound graph output

fgmacedo committed:
fix: preserve initial states order from SCXML declaration

The SCXML `initial` attribute specifies states in declaration order
(e.g., `initial="s2p112 s2p122"`), but `_parse_initial` returned a
`set`, losing ordering due to hash randomization. This caused flaky
test413 — the parallel state's children were entered in random order,
sometimes triggering eventless transitions to `fail` before both
children were in the configuration.

Change `initial_states` from `Set[str]` to `List[str]` in
`StateMachineDefinition` and `_parse_initial`, preserving declaration
order for `start_configuration_values`. The `all_initial_states` set
is still used for O(1) membership checks in `parse_state`.

fgmacedo committed:
fix: set ruff target-version to py39 (minimum supported version)

With target-version py314, ruff format applies PEP 758 (parentheses-free
except clauses), which is invalid syntax on Python < 3.14. Since the
project supports Python 3.9+, the target must match the minimum version
to prevent incompatible syntax in formatted code.

fgmacedo committed:
Merge branch 'develop' into macedo/scxml

# Conflicts:
#	pyproject.toml
#	uv.lock

fgmacedo committed:
feat: implement error.execution event handling per SCXML spec

Add `error_on_execution` flag (True on StateChart, False on StateMachine
for backward compat) that catches runtime exceptions in callbacks and
queues an `error.execution` internal event instead of propagating.

Error handling is layered by responsibility:
- callbacks.call(on_error=...): per-block isolation for onentry/onexit
  — errors in one block don't affect other blocks
- callbacks.all(on_error=...): condition errors treated as False
- Engine._conditions_match: delegates to callbacks.all with on_error
- Engine.microstep: catches errors in transition actions with rollback
- SCXML if_action: condition errors treated as false (SCXML translation)
- InvalidDefinition always propagates regardless of error_on_execution

Also fixes:
- Event with custom id in factory (preserve SCXML event ids like
  "error.execution" while keeping Python attribute accessible)
- SCXML _event system variable initialized to None before first event
- Infinite loop guard for error.execution in transition handlers

fgmacedo committed:
feat: add error_ naming convention, statecharts docs, and error handling tests

Auto-register dot-notation aliases for event attributes starting with
`error_` (e.g. `error_execution` matches both `error_execution` and
`error.execution`), removing the need for verbose `id="error.execution"`.

Add `docs/statecharts.md` documenting StateChart vs StateMachine class-level
flags (allow_event_without_transition, enable_self_transition_entries,
atomic_configuration_update, error_on_execution) and error.execution handling.

Comprehensive test coverage for error handling with conditions, listeners,
flow control, and SCXML spec compliance using LOTR-themed scenarios.

fgmacedo committed:
feat: implement SCXML donedata support for final states

Parse <donedata> elements (<param> and <content>) from <final> states,
evaluate expressions at runtime, and pass the data correctly to
done.state.* events. Error cases (invalid expr/location) raise
error.execution per SCXML spec.

Fixes W3C conformance tests: 294, 298, 343, 488, 527, 528, 529.

fgmacedo committed:
test: improve test coverage to 99% across statemachine package

Add comprehensive tests organized by scope in existing test files:
- Async engine: error propagation, InvalidDefinition handling, start noop
- Error execution: internal event error handling, microstep error paths
- State machine: compound state configuration, abstract SM, raise_ events
- Diagrams: compound/parallel subgraphs, dashed style, lhead attributes
- SCXML units: parser errors, action callables, schema edge cases
- Transitions: non-callable call, loop transitions, event name tests

Mark confirmed dead code with pragma no cover:
- factory.py: abstract class checks unreachable after early return
- sync.py: _run_microstep safety net (microstep handles errors internally)
- sync.py: secondary internal queue drain (macrostep loop already drains)
- base.py: is_in_final_state parallel/atomic branches (nested parallel)

fgmacedo committed:
feat: support multi-target transitions in Transition class

The `target` parameter now accepts `State | List[State] | None`,
enabling SCXML parallel region entry where a single transition
targets multiple states. Internally stores as `_targets` list with
backward-compatible `target` property (returns first target) and
new `targets` property (returns full list).

fgmacedo committed:
feat: async engine parity, multi-target transitions, and SCXML send fixes

- Rewrite AsyncEngine to mirror SyncEngine's 3-phase macrostep
  architecture with full StateChart support
- Extract pure computation helpers from BaseEngine so both engines
  share the same logic
- Add on_error callback support to async dispatch methods
- Support multi-target initial transitions for SCXML parallel regions
  in engine, IO layer, parser, and factory
- Fix SCXML send error handling: error.communication for undispatchable
  session targets, error.execution for invalid targets/types, and
  namelist variable validation
- Remove resolved xfail markers (test364, test496, test521, test553)

fgmacedo committed:
refactor: move donedata support from SCXML module to core State

Add `donedata` parameter to `State.__init__()` so Python-defined final
states can specify donedata callables directly. The SCXML processor now
passes donedata as a separate key instead of mixing it into enter
callbacks. Also removes the no-op `_processing_loop()` call from
`ExecuteBlock`.

fgmacedo committed:
feat: run SCXML test suite with both sync and async engines

Add async SCXML test variant using a minimal AsyncListener to trigger
AsyncEngine selection, keeping SCXMLProcessor free of async concerns.
Fix __initial__ handling in AsyncEngine (break instead of continue in
phase 3) to ensure internal events are processed before external ones.

fgmacedo committed:
feat: diagram generation for compound, parallel, and history states

- Render initial pseudo-state (black dot → initial child) inside all
  compound subgraphs, not just root level
- Add history state rendering as UML circles labeled "H" / "H*"
- Annotate parallel state subgraph labels with ☷ indicator
- Support multi-target transitions (one edge per target)
- Extract _add_transitions() helper to reduce _graph_states() complexity
- Remove stale TODO comment
- Fix NestedStateFactory to propagate kwargs (e.g. parallel=True) from
  State.Parallel/State.Compound base classes to subclass-created States
- Export HistoryState from statemachine.__init__
- Register event label "None" bug as release blocker in PLAN

fgmacedo committed:
tests: Trying to figure out error on CI

fgmacedo committed:
refactor: unify error.execution handling in engines

Replace scattered `if self.sm.error_on_execution` checks and the broken
`_on_error_execution` property with two unified methods:

- `_on_error_handler(trigger_data)`: returns a bound callable (or None)
  for per-block callback error isolation (onentry/onexit/conditions).
- `_handle_error(error, trigger_data)`: for try/except blocks in
  microstep/_run_microstep — sends error.execution or re-raises.

Fixes two bugs in the previous code:
- base.py used `partial(self._on_error_execution, trigger_data)` which
  bound trigger_data to the wrong positional parameter (error), causing
  AttributeError on any per-block callback error.
- async_.py passed `self._on_error_execution` without binding
  trigger_data at all, which would fail with a missing argument.

Also fixes `_send_error_execution` parameter order to (error,
trigger_data) and removes the unused `functools.partial` import.

fgmacedo committed:
Merge remote-tracking branch 'origin/develop' into macedo/scxml

fgmacedo committed:
fix: handle HistoryState separately in NestedStateFactory

NestedStateFactory.__new__ was treating HistoryState as regular State
(via isinstance) and adding it to states=[]. HistoryState instances
need to go into history=[] for proper initialization in State.__init__.

fgmacedo committed:
test: add Python-syntax StateChart tests (55 tests)

Comprehensive test coverage for StateChart features using Python class
syntax: compound states, parallel states, history states, eventless
transitions, donedata, delayed events, error.execution, and In() conditions.

fgmacedo committed:
fix: sort imports in test_signature.py

fgmacedo committed:
test: run StateChart tests on both sync and async engines

Add SMRunner fixture parametrized over sync/async to conftest.py.
Engine-dependent tests (53) now run on both engines via sm_runner,
producing 108 test cases total. Definition-only tests (2) remain sync.

fgmacedo committed:
feat: add done_state_ naming convention and complete StateChart docs

Add done_state_ prefix handling in factory.py alongside the existing
error_ convention. Attributes starting with done_state_ auto-register
both the underscore and dot forms (e.g., done_state_quest matches
both "done_state_quest" and "done.state.quest"). Only the prefix is
replaced, preserving multi-word state names (done_state_lonely_mountain
→ done.state.lonely_mountain).

Simplify existing tests, examples, and docs to use the convention
instead of explicit id= parameters.

Add comprehensive StateChart documentation: compound states, parallel
states, history pseudo-states, eventless transitions, DoneData,
delayed events, In() conditions, error.execution, and the new
done_state_ convention.

fgmacedo committed:
fix: None event label in create_machine_class_from_definition + docs

Fix bug where transition_data.get("event") returning None was
interpolated as literal "None" in event IDs (e.g., "next None").

Complete remaining documentation: configuration property in states.md,
cross-boundary transitions and transition priority in transitions.md,
async-specific limitations in async.md, prepare_event convention and
create_machine_class_from_definition docstring in api.md.

fgmacedo committed:
merge: integrate develop (v2.6.0) into macedo/scxml

Resolved conflicts in engines (adapted enabled_events for StateChart
architecture using configuration instead of current_state), statemachine.py,
transition_mixin.py (took develop's TypeError/decorator-syntax bugfix),
pyproject.toml (merged python_version markers with xdist/timeout deps),
and test files (kept both scxml and enabled_events tests).

Also restored engine.start() call in __setstate__ for async copy support.

fgmacedo committed:
fix: address SonarCloud code smells

- S1186: add comments to empty method bodies in tests
- S125: remove commented-out code in conftest.py and test_microwave.py
- S1192: extract _ERROR_EXECUTION constant in engines/base.py
- S108: invert condition to eliminate empty pass block in base engine
- S5713: remove redundant IndentationError (subclass of SyntaxError)
- S1940: use != instead of not == in processor.py
- S2772: remove unneeded pass in schema.py Action dataclass
- S1854: remove dead assignment to all_initial_states in parser.py
- S5806: rename id -> func_id to avoid shadowing builtin in spec_parser.py
- S1172: remove unused is_root parameter from diagram._graph_states
- S5655: fix Events.match type hint to accept str | None
- S5886: fix return type of create_machine_class_from_definition

fgmacedo committed:
test: achieve 100% test coverage

Add tests and pragmas to eliminate all coverage gaps:

- Add tests for enabled_events dedup, history diagram node, IO
  definition parsing (_parse_history, event name concatenation),
  and SCXML history without transitions
- Remove unused trigger_data param from _on_error_handler
- Remove unused Transition._get_event_by_id method
- Mark genuinely unreachable defensive code with pragma comments
  (callbacks on_error path, SCXML algorithm guards, factory parallel
  initial transitions, parser/processor safety checks)
